data/generate_embeddings.py

```diff
@@ -0,0 +1,598 @@
+
+    # Try exact absolute path
+    city = path_to_city.get(abs_posix)
+    if city:
+        return city
+
+    # Try relative to project root
+    city = path_to_city.get(Path(rel_from_cwd).as_posix())
+    if city:
+        return city
+
+    # Fallback to basename if unique in metadata
+    base = input_path.name
+    city = basename_to_city.get(base)
+    if city:
+        return city
+
+    return "unknown"
+
+
+def ensure_output_dir(output_dir: Path):
+    """Create output directory if it doesn't exist."""
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+
+def get_next_global_chunk_id(output_dir: Path) -> int:
+    """Return the next global sequential chunk ID based on existing files.
+
+    Scans filenames like '<city>_NNN.json' and returns max(NNN)+1.
+    Note: This is O(n) over files in the output directory and is sufficient
+    for moderate corpus sizes. For very large directories, consider tracking
+    the last ID in a small sidecar file (e.g., 'index.json').
+    """
+    max_id = 0
+    for file_path in output_dir.glob("*.json"):
+        stem = file_path.stem
+        if "_" not in stem:
+            continue
+        id_part = stem.split("_")[-1]
+        if id_part.isdigit():
+            try:
+                max_id = max(max_id, int(id_part))
+            except ValueError:
+                continue
+    return max_id + 1
+
+# -------------------------
+# Embedding Provider
+# -------------------------
+class EmbeddingProvider:
+    """Handles OpenAI embedding API calls."""
+    
+    def __init__(self, provider: str, model: str, api_key: str = None):
+        self.provider = provider
+        self.model = model
+        self._client = None
+
+        try:
+            if not api_key:
+                api_key = os.getenv("OPENAI_API_KEY")
+            if not api_key:
+                raise RuntimeError("OpenAI API key must be provided in OPENAI_API_KEY environment variable")
+            self._client = OpenAI(api_key=api_key)
+        except ImportError:
+            raise RuntimeError("OpenAI client not available. Install with: pip install openai")
+
+    def embed_batch(self, texts: List[str], retry_attempts: int = None, retry_min_wait: int = None, retry_max_wait: int = None) -> List[List[float]]:
+        """Generate embeddings for a batch of texts with retry logic."""
+        if not TENACITY_AVAILABLE:
+            # Simple retry without tenacity
+            return self._embed_batch_simple(texts)
+        
+        # Use provided values or defaults
+        attempts = retry_attempts or DEFAULT_RETRY_ATTEMPTS
+        min_wait = retry_min_wait or DEFAULT_RETRY_MIN_WAIT
+        max_wait = retry_max_wait or DEFAULT_RETRY_MAX_WAIT
+        
+        @retry(
+            reraise=True,
+            stop=stop_after_attempt(attempts),
+            wait=wait_exponential(multiplier=1, min=min_wait, max=max_wait),
+            retry=retry_if_exception_type(Exception),
+        )
```

Comment:
If `TENACITY_AVAILABLE` is not set, consider a simple exponential backoff instead of immediate failover.
Log API errors with more details, including batch size and input length.